embassy-imxrt/src/flexcomm/uart.rs

```diff
@@ -0,0 +1,1220 @@
+        }
+    }
+}
+
+impl<'a, M: Mode> UartRx<'a, M> {
+    fn new_inner<T: Instance>(rx_dma: Option<Peri<'a, AnyChannel>>) -> Self {
+        let uartrx = Self {
+            info: T::info(),
+            rx_dma,
+            _phantom: PhantomData,
+        };
+        uartrx.info.refcnt.fetch_add(1, Ordering::Relaxed);
+        uartrx
+    }
+}
+
+impl<'a, M: Mode> Drop for UartRx<'a, M> {
+    fn drop(&mut self) {
+        if self.info.refcnt.fetch_sub(1, Ordering::Relaxed) == 1 {
+            while self.info.regs.stat().read().rxidle().bit_is_clear() {}
+
+            self.info.regs.fifointenclr().modify(|_, w| {
+                w.txerr()
+                    .set_bit()
+                    .rxerr()
+                    .set_bit()
+                    .txlvl()
+                    .set_bit()
+                    .rxlvl()
+                    .set_bit()
+            });
+
+            self.info
+                .regs
+                .fifocfg()
+                .modify(|_, w| w.dmatx().clear_bit().dmarx().clear_bit());
+
+            self.info.regs.cfg().modify(|_, w| w.enable().disabled());
+        }
+    }
+}
+
+impl<'a> UartRx<'a, Blocking> {
+    /// Create a new blocking UART which can only receive data
+    pub fn new_blocking<T: Instance>(_inner: Peri<'a, T>, rx: Peri<'a, impl RxPin<T>>, config: Config) -> Result<Self> {
+        rx.as_rx();
+
+        let _rx = rx.into();
+        Uart::<Blocking>::init::<T>(None, Some(_rx), None, None, config)?;
+
+        Ok(Self::new_inner::<T>(None))
+    }
+}
+
+impl UartRx<'_, Blocking> {
+    fn read_byte_internal(&mut self) -> Result<u8> {
+        if self.info.regs.fifostat().read().rxerr().bit_is_set() {
+            self.info.regs.fifocfg().modify(|_, w| w.emptyrx().set_bit());
+            self.info.regs.fifostat().modify(|_, w| w.rxerr().set_bit());
+            Err(Error::Read)
+        } else if self.info.regs.stat().read().parityerrint().bit_is_set() {
+            self.info.regs.stat().modify(|_, w| w.parityerrint().clear_bit_by_one());
+            Err(Error::Parity)
+        } else if self.info.regs.stat().read().framerrint().bit_is_set() {
+            self.info.regs.stat().modify(|_, w| w.framerrint().clear_bit_by_one());
+            Err(Error::Framing)
+        } else if self.info.regs.stat().read().rxnoiseint().bit_is_set() {
+            self.info.regs.stat().modify(|_, w| w.rxnoiseint().clear_bit_by_one());
+            Err(Error::Noise)
+        } else {
+            let byte = self.info.regs.fiford().read().rxdata().bits() as u8;
+            Ok(byte)
+        }
+    }
+
+    fn read_byte(&mut self) -> Result<u8> {
+        if self.info.regs.fifostat().read().rxnotempty().bit_is_clear() {
+            Err(Error::RxFifoEmpty)
+        } else {
+            self.read_byte_internal()
+        }
+    }
+
+    fn blocking_read_byte(&mut self) -> Result<u8> {
+        while self.info.regs.fifostat().read().rxnotempty().bit_is_clear() {}
+        self.read_byte_internal()
```

Comment:
Should there be a compiler fence here to forbid the compiler from reading the from the hardware fifo before the fifo actually contains anything?

Reply:
I would think maybe the volatile reads would keep ordering consistent, but I'd prefer to be explicit.